README.md
# pilotedge_helper
X-Plane 10 LUA script with some handy functions for online flying on Pilotedge, Vatsim, etc. Requires [FlyWithLUA plugin](http://forums.x-plane.org/index.php?app=downloads&showfile=17468) installed

## Features

### Mini transponder gauge

An unobtrusive mouse-driven transponder gauge, giving you the ability to always have a "mode C" capable transponder with IDENT functionality on any airplane/model that you use.

The gauge is hidden by default and activated by moving your mouse to the **bottom right corner** of your monitor: the gauge then appears and you can **change your squawk code** by mousing over the individual digits and rotating the mouse wheel up or down.

![Gauge](http://i.imgur.com/vK4mN2R.jpg)

The colored vertical bar on the left represents the **mode of the transponder**:

* red = transponder off
* yellow = transponder standby
* green = transponder on/alt (mode "C")

![SQ SBY](http://i.imgur.com/fPh3kBh.jpg)
![SQ C](http://i.imgur.com/MfWogE7.jpg)

The **modes can be cycled** by left-clicking anywhere within the main transponder window: click once to go from OFF to SBY, then one more time to go from SBY to ON/ALT the once more to go from ON to OFF, etc.

*NOTE: Please note that neither X-Plane nor the network pilot clients distinguish between the modes ON and ALT. Some airplane models do have transponders that visually represent these different modes, but that's purely cosmetical and uses the same internal X-Palne dataref: as far as X-Plane and the pilot clients are concerned there are three transponder states only (+ a test state which is interpreted as ON by the clients as well) and when the controller tells you to "squawk C" or "squawk altitude" if you're green, you're good*

The upper part of the gauge represents the **IDENT button**: clicking anywhere in that upper area will send the ident signal and this is what you press when a controller asks you to ident. While identing, the upper area will blink very fast in green color, to give you a visual cue that you're identing.

![SQ IDENT](http://i.imgur.com/RZ1Lc8x.jpg)

Moving your mouse outside of the gauge area hides the gauge, so you can enjoy an uncluttered screen without breaking your immersion.

*NOTE: The gauge is fully compatible with plane models that already have built-in transponders: in these cases it can be treated just as an alternate method of manipulating the transponder values, or can be completely ignored (or disabled: see below)*

### FPS counter

On the bottom left of your screen you'll see several indicators, the first of which is a FPS counter:

![FPS](http://i.imgur.com/AsanbRW.jpg)

True, you can opt to visualize your FPS by displaying the relevant X-Plane dataref in the upper left corner, but it's a big ugly box that holds a ton of additional information that you probably don't need unless you're troubleshooting something.
This here is a simple **color coded** FPS counter: with frames at 30 or above it's green, the range 20 to 30 is yellow and at 19 or bellow it goes red.

*NOTE: The 19 FPS threshold is an important one when flying online: when you drop below it X-Plane will start transparently decreasing your simulation rate, so that your system is able to cope with the rendering strain. The net result is that you will be moving around more slowly than you think you are (and than your gauges are telling you). For the controllers watching you on their scopes and for other pilots in the air your ground speed would be slower than displayed: e.g. you might be at 120kt indicated / 140 ground speed according to your speedometer, but due to the sim-rate decrease you might be effectively at let's say 100 ground speed and this is how it would appear to the other guys on the network, be it ATC or other pilots. You can ocassionally hear controllers on Pilotedge commenting on this on the frequency and mentioning "the FPS issue"*

### Selected COM radio (and frequency)

The second indicator shows you the currently selected COM radio **for transmitting** (along with its currently dialed frequency):

![COM](http://i.imgur.com/h4x9VTj.jpg)

When flying online and having to deal with multiple frequency changes it's convenient to use both COM radios (when your plane has them). Unfortunately many models / addons fall short when it comes to selecting the radio for transmitting: some addons just don't have working buttons for that in their 3D cockpits, some tie it to the listening radio selections, some use custom code, etc. It's inconsistent and I've heard pilots on Pilotedge many times unintentionally transmit on the wrong COM radio (and I've done it myself).
So it's very handy to be able to *determine at a glance* which COM radio is selected for transmission and what frequency is dialed in it. This is what this indicator does and it's also *color coded*: it's white when COM1 is selected, yellow when COM2 is selected and red while you're transmitting (the "Contact ATC" button pressed):

![COM2](http://i.imgur.com/P6kgzyN.jpg)
![COM TRANSMIT](http://i.imgur.com/zuRuYsT.jpg)

Incoming transmissions are also monitored and represented visually: while a transmission is being received, the radio text will be wrapped in ">...<" symbols.

Apart from being an indicator, this area is also an actuator, i.e. it not only shows you the information, but allows you to **change it (COM1 <-> COM2)**: clicking anywhere on the COM area will toggle the transmitting radio from COM1 to COM2 or vice versa. You can use that if your plane/hardware does not have a convenient way of switching between the radios.

### Transponder mode indicator

The last indicator displays the current transponder mode, which is also handy to have at a glance at all times: you ought to have your transponder on when airborne at sometimes also on the ground when the airport uses Ground Surveilance Radar. The indicator is **color coded** and displays a red [-] symbol when transponder is off, yellow [S] when transponder is on standby and green [C] when transponder is on (mode C):

![SQ MODE](http://i.imgur.com/b6POU0U.jpg)

### Electrical system fixes for various aircraft addons

The script includes dynamic fixes to the electrical systems of some addons, in order to correct unrealistic values, dummy switches or other deficiencies and to avoid loss of comms due to unwarranted battery discharge.

The following addons are supported:

#### SMS DHC-2 Beaver

* Coupled the dummy/custom alternator switch with X-Plane's generator dataref to avoid gradual battery discharge and comms loss during longer flights. Engaging or disengaging the plane's alternator will now correctly pass the event to X-Plane's built-in systems.

## Installation

Firstly you need to have the  [FlyWithLUA plugin](http://forums.x-plane.org/index.php?app=downloads&showfile=17468) installed. You might already have it if you're using RealTerraHaze or some of the other popular scripts that require it.

To get the PIlotedge Helper Script go to: https://github.com/tarakanbg/pilotedge_helper and press the "Download ZIP" button:

![Download](http://i.imgur.com/RrFT0ha.png)

Unpack the ZIP archive, grab the `pilotedge_helper.lua` file and drop it in your X-Plane `\Resources\plugins\FlyWithLua\Scripts` folder.  That's it.

## Customization

You can optionally edit the first few lines of the script in your preferred text editor to customize it:

![Customize](http://i.imgur.com/4yqNHR8.png)

The preferences are described in the file and should be self-explanatory: you write `1` or `0` to enable/disable individual components depending on your needs.

Enjoy!


CHANGELOG.md
26 May 2016
-----------
* Fixed a bug with COM khz freq being unwritable when using the tuning knobs of the GNS.
* Fixed a bug with the color of the TX indicator


23 March 2016
--------------
* added monitoring and depiction of Pilotedge RX dataref for identifying transmissions being received on the active radio
* added dynamic fixes for the SMS DHC-2 Beaver electrical system: coupled the dummy alternator switch with X-Plane's generator
  dataref to avoid gradual battery discharge and comms loss during longer flights


09 March 2016
--------------

* added native monitoring of Pilotedge TX dataref for identifying active transmissions


pilotedge_helper.lua
-- PILOTEDGE helper
-- Simple monitor and transponder to enhance online flying
-- (c) Svilen Vassilev
-- build 2016-05-26
-- Documentation (manual) and latest version at: https://github.com/tarakanbg/pilotedge_helper

--| ----------------------------------|
--|  Customize your preferences here  |
--| ----------------------------------|

-- Completely disables all visual output. Use this to disable 
-- the script without removing it from your "Scripts" folder.
disable_data_display = 0

-- Toggle FPS counter. 1 = Enabled / 0 = Disabled
show_fps = 1

-- Toggle COM radio display. 1 = Enabled / 0 = Disabled
-- Shows the COM radio selected for TRANSMITTING
show_COM_and_frequency = 1

-- Toggle squawk mode display. 1 = Enabled / 0 = Disabled
show_squawk_mode = 1

-- Toggle transponder mini gauge. 1 = Enabled / 0 = Disabled
enable_transponder_minigauge = 1

-- Toggle electrical system fixes for various aircraft models:

-- When set to 1 completely disables all electrical fixes for all aircraft (default 0)
disable_electrical_fixes = 0

-- Enables the coupling of the SMS DHC-2 Beaver alternator switch with
-- X-plane's generator dataref to avoid gradual battery discharge with
-- alternator "on"
enable_sms_dhc2_alternator_fix = 1

--_________________  End of preferences  _________________________--

--________________________________________________________--

--| ----------------------------------|
--|  !! DO NOT EDIT PAST THIS LINE !! |
--| ----------------------------------|

--________________________________________________________--

if disable_electrical_fixes== 0 then

  -- function used for testing
  -- do_every_draw("show_ICAO()")
  -- function show_ICAO()
  --   draw_string(10, 500, AIRCRAFT_FILENAME)
  -- end

  if enable_sms_dhc2_alternator_fix==1 then
    if AIRCRAFT_FILENAME == "SMS_Beaver_Regular.acf" or AIRCRAFT_FILENAME == "SMS_Beaver_Floats.acf" or AIRCRAFT_FILENAME == "SMS_Beaver_Amphib.acf" then
      do_often("determine_dhc2_alternator()")
    end
  end

end

if disable_data_display==0 then

  -- require "radio"

  if show_fps==1 then
    do_every_frame("FPS_count()")
    do_often("FPS_capture()")
    do_every_draw("FPS_draw()")
  end

  if show_COM_and_frequency==1 then
    do_every_frame("determine_radio()")
    do_every_draw("Radio_draw()")
    do_on_mouse_click("radio_mouse_click_events()")
  end

  if show_squawk_mode==1 then
    do_every_frame("determine_transponder()")
    do_every_draw("sq_draw()")
  end

  if enable_transponder_minigauge==1 then
    require "graphics"
    do_every_draw("draw_transponder_gauge()")
    do_on_mouse_click("transponder_mouse_click_events()")
    do_on_mouse_wheel("transponder_wheel_events()")
  end

  fps_pos_x = 5
  fps_pos_y = 5

  radio_position_x = 55
end

--________________________________________________________--
-- FPS counter --
--________________________________________________________--

fps=-1
fps_cnt=0

function FPS_count()
  fps_cnt=fps_cnt+1
end

function FPS_capture()
  fps=fps_cnt
  fps_cnt=0
end

function FPS_draw()
  if(fps>=0 and show_fps==1) then
    if fps<=19  then
      draw_string(fps_pos_x, fps_pos_y, "FPS "..fps, "red")
    elseif fps>=30 then
      draw_string(fps_pos_x, fps_pos_y, "FPS "..fps, "green")
    else
      draw_string(fps_pos_x, fps_pos_y, "FPS "..fps, "yellow")
    end
  end
end

--________________________________________________________--
-- RADIO display --
--________________________________________________________-


function determine_radio()
  dataref("COM1", "sim/cockpit2/radios/actuators/com1_frequency_hz")
  dataref("COM2", "sim/cockpit2/radios/actuators/com2_frequency_hz")
  com1_mhz = string.sub(COM1, 1, 3)
  com1_khz = string.sub(COM1, 4, 5)
  com1_humanized = com1_mhz.."."..com1_khz

  com2_mhz = string.sub(COM2, 1, 3)
  com2_khz = string.sub(COM2, 4, 5)
  com2_humanized = com2_mhz.."."..com2_khz


  dataref("selected_radio", "sim/cockpit2/radios/actuators/audio_com_selection")

  if XPLMFindDataRef("pilotedge/radio/tx_status") ~= nil then
    dataref("tx_on", "pilotedge/radio/tx_status")
  end
  if XPLMFindDataRef("pilotedge/radio/rx_status") ~= nil then
    dataref("rx_on", "pilotedge/radio/rx_status")
  end

  if selected_radio==6 then
    if rx_on == 1 then
      radio_string = "> COM1: "..com1_humanized.." <"
    else
      radio_string = "COM1: "..com1_humanized
    end
    radio_color = "white"
  elseif selected_radio==7 then
    if rx_on == 1 then
      radio_string = "> COM2: "..com2_humanized.." <"
    else
      radio_string = "COM2: "..com2_humanized
    end
    radio_color = "yellow"
  else
    radio_string = "NO RADIO"
    radio_color = "red"
  end
  if tx_on == 1 then
    radio_color = "red"
  end

end

function Radio_draw()
  draw_string(radio_position_x, 5, radio_string, radio_color)
end

function radio_mouse_click_events()
  if MOUSE_STATUS ~= "down" then
    return
  end

  if (MOUSE_X > 55 and MOUSE_X < 135) and (MOUSE_Y > 4 and MOUSE_Y < 30) then
    dataref("transmit_radio", "sim/cockpit2/radios/actuators/audio_com_selection", "writable")
    if transmit_radio==6 then
      transmit_radio = 7
    elseif transmit_radio==7 then
      transmit_radio = 6
    else
      transmit_radio = 6
    end
  end
end

--________________________________________________________--
-- TRANSPONDER mode --
--________________________________________________________-

function determine_transponder()

  dataref("TRANSPONDER_MODE", "sim/cockpit/radios/transponder_mode", "writable")

  if TRANSPONDER_MODE == 0 then
    transponder_color = "red"
    transponder_on = 0
    transponder_string = "-"
  elseif TRANSPONDER_MODE == 1 then
    transponder_color = "yellow"
    transponder_on = 0
    transponder_string = "S"
  else
    transponder_color = "green"
    transponder_on = 1
    transponder_string = "C"
  end
end

function sq_draw()
  if rx_on and rx_on ==1 then
    pos_trans = 160
  else
    pos_trans = 140
  end
  draw_string(pos_trans, 5, "[ "..transponder_string.." ]", transponder_color)
end

--________________________________________________________--
-- TRANSPONDER mini gauge --
--________________________________________________________-

function draw_transponder_gauge()

  dataref("TRANSPONDER_MODE", "sim/cockpit/radios/transponder_mode", "writable")
  dataref("SQUAWK", "sim/cockpit/radios/transponder_code", "writable")

  -- does we have to draw anything?
  if MOUSE_Y > 80 or MOUSE_X < SCREEN_WIDTH - 100 then
    return
  end

  -- init the graphics system
  XPLMSetGraphicsState(0,0,0,1,1,0,0)

  -- draw transparent backgroud
  graphics.set_color(0, 0, 0, 0.5)
  graphics.draw_rectangle(SCREEN_WIDTH - 100, 0, SCREEN_WIDTH, 50)


  -- draw colored switches info
  if TRANSPONDER_MODE == 0 then
    graphics.set_color(1, 0, 0, 0.5)
  elseif TRANSPONDER_MODE == 1 then
    graphics.set_color(1, 1, 0, 0.5)
  else
    graphics.set_color(0, 1, 0, 0.5)
  end
  graphics.draw_rectangle(SCREEN_WIDTH - 100, 0, SCREEN_WIDTH - 92, 50)

  -- ID window
  dataref("IDENT", "sim/cockpit2/radios/indicators/transponder_id")
  if IDENT == 0 then
    graphics.set_color(0, 0, 0, 0.5)
  else
    math.randomseed(os.clock()*100000000000)
    ran = math.random(1, 2)
    if ran == 1 then
      graphics.set_color(0, 1, 0, 0.5)
    else
      graphics.set_color(0, 1, 0, 1)
    end
  end
  graphics.draw_rectangle(SCREEN_WIDTH - 100, 50, SCREEN_WIDTH, 75)
  graphics.set_color(1, 1, 1, 1)
  graphics.set_width(2)
  graphics.draw_line(SCREEN_WIDTH - 100, 51, SCREEN_WIDTH - 100, 75)
  graphics.draw_line(SCREEN_WIDTH - 100, 75, SCREEN_WIDTH - 1, 75)
  graphics.draw_line(SCREEN_WIDTH - 1, 75, SCREEN_WIDTH - 1, 51)
  graphics.draw_line(SCREEN_WIDTH, 51, SCREEN_WIDTH - 100, 51)

  -- draw lines around the hole block
  graphics.set_color(1, 1, 1, 1)
  graphics.set_width(2)
  graphics.draw_line(SCREEN_WIDTH - 100, 1, SCREEN_WIDTH - 100, 50)
  graphics.draw_line(SCREEN_WIDTH - 100, 50, SCREEN_WIDTH - 1, 50)
  graphics.draw_line(SCREEN_WIDTH - 1, 50, SCREEN_WIDTH - 1, 1)
  graphics.draw_line(SCREEN_WIDTH, 1, SCREEN_WIDTH - 100, 1)

  -- draw the info text
  draw_string_Helvetica_10(SCREEN_WIDTH - 80, 38, "SQUAWK")
  draw_string_Helvetica_12(SCREEN_WIDTH - 70, 60, "IDENT")

  sq_len = string.len (SQUAWK)

  if sq_len == 1 then
    digit1 = 0
    digit2 = 0
    digit3 = 0
    digit4 = string.sub(SQUAWK, 1, 1)
  elseif sq_len == 2 then
    digit1 = 0
    digit2 = 0
    digit3 = string.sub(SQUAWK, 1, 1)
    digit4 = string.sub(SQUAWK, 2, 2)
  elseif sq_len == 3 then
    digit1 = 0
    digit2 = string.sub(SQUAWK, 1, 1)
    digit3 = string.sub(SQUAWK, 2, 2)
    digit4 = string.sub(SQUAWK, 3, 3)
  elseif sq_len == 4 then
    digit1 = string.sub(SQUAWK, 1, 1)
    digit2 = string.sub(SQUAWK, 2, 2)
    digit3 = string.sub(SQUAWK, 3, 3)
    digit4 = string.sub(SQUAWK, 4, 4)
  else
    digit1 = 7
    digit2 = 7
    digit3 = 7
  digit4 = 7
  end
  draw_string_Times_Roman_24(SCREEN_WIDTH - 85, 15, digit1)
  draw_string_Times_Roman_24(SCREEN_WIDTH - 65, 15, digit2)
  draw_string_Times_Roman_24(SCREEN_WIDTH - 45, 15, digit3)
  draw_string_Times_Roman_24(SCREEN_WIDTH - 25, 15, digit4)
end

function transponder_mouse_click_events()

  dataref("TRANSPONDER_MODE", "sim/cockpit/radios/transponder_mode", "writable")

  -- we will only react once
  if MOUSE_STATUS ~= "down" then
    return
  end

  if MOUSE_X > SCREEN_WIDTH - 100 and MOUSE_Y < 50 then
    if TRANSPONDER_MODE == 0 then
      TRANSPONDER_MODE = 1
    elseif TRANSPONDER_MODE == 1 then
      TRANSPONDER_MODE = 2
    else
      TRANSPONDER_MODE = 0
    end
    RESUME_MOUSE_CLICK = true
  end

  if MOUSE_X > SCREEN_WIDTH - 100 and (MOUSE_Y > 50 and MOUSE_Y < 75) then
    command_once("sim/transponder/transponder_ident")
  end
end

function transponder_wheel_events()
  dataref("SQUAWK", "sim/cockpit/radios/transponder_code", "writable")

  if MOUSE_Y > 50 or MOUSE_X < SCREEN_WIDTH - 100 then
    return
  end

  --digit 1
  if MOUSE_X > SCREEN_WIDTH - 90 and MOUSE_X < SCREEN_WIDTH - 70 and MOUSE_Y < 50 then
    digit1 = digit1 + MOUSE_WHEEL_CLICKS
    if digit1 > 7 then
      digit1 = 7
    end
    if digit1 < 0 then
      digit1 = 0
    end
  end

  --digit 2
  if MOUSE_X > SCREEN_WIDTH - 70 and MOUSE_X < SCREEN_WIDTH - 50 and MOUSE_Y < 50 then
    digit2 = digit2 + MOUSE_WHEEL_CLICKS
    if digit2 > 7 then
      digit2 = 7
    end
    if digit2 < 0 then
      digit2 = 0
    end
  end

  --digit 3
  if MOUSE_X > SCREEN_WIDTH - 50 and MOUSE_X < SCREEN_WIDTH - 30 and MOUSE_Y < 50 then
    digit3 = digit3 + MOUSE_WHEEL_CLICKS
    if digit3 > 7 then
      digit3 = 7
    end
    if digit3 < 0 then
      digit3 = 0
    end
  end

  --digit 4
  if MOUSE_X > SCREEN_WIDTH - 30 and MOUSE_X < SCREEN_WIDTH - 10 and MOUSE_Y < 50 then
    digit4 = digit4 + MOUSE_WHEEL_CLICKS
    if digit4 > 7 then
      digit4 = 7
    end
    if digit4 < 0 then
      digit4 = 0
    end
  end
  SQUAWK = digit1..digit2..digit3..digit4
  RESUME_MOUSE_WHEEL = true
end

--________________________________________________________--
-- Electrical system fixes --
--________________________________________________________-

-- SMS DHC-2 Beaver Alternator Fix --

function determine_dhc2_alternator()
  if AIRCRAFT_FILENAME == "SMS_Beaver_Regular.acf" or AIRCRAFT_FILENAME == "SMS_Beaver_Floats.acf" or AIRCRAFT_FILENAME == "SMS_Beaver_Amphib.acf" then
    dataref("icao", "sim/aircraft/view/acf_ICAO")
    if XPLMFindDataRef("sms/custom/cdrControlsAlternator") ~= nil then
      dataref("dhc2_alternator", "sms/custom/cdrControlsAlternator")
      dataref("xplane_generator", "sim/cockpit/electrical/generator_on", "writable")
      if dhc2_alternator == 1 then
        xplane_generator = 1
      else
        xplane_generator = 0
      end
    end
  end
end
